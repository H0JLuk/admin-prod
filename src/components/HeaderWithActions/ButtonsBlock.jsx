import React from 'react';
import { Button } from 'antd';

const ButtonsBlock = ({ buttons, params }) => {
    return (buttons || []).map((button) => (
        <Button
            key={ button.label }
            type={ button.type }
            onClick={ button.onClick }
            disabled={ typeof button.disabled === 'function' ? button.disabled(params) : button.disabled }
        >
            { button.label }
        </Button>
    ));
};

export default ButtonsBlock;
