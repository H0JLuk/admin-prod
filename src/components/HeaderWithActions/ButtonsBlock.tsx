import React from 'react';
import { Button, ButtonProps as AntButtonProps } from 'antd';

export type ButtonProps = Omit<AntButtonProps, 'disabled' | 'label'> & {
    label: string;
    disabled?: ((params?: Record<string, string | number>) => boolean) | boolean;
};

type ButtonsBlockProps = {
    buttons: ButtonProps[];
    params?: Record<string, string | number>;
};

const ButtonsBlock: React.FC<ButtonsBlockProps> = ({ buttons, params }) =>
    (buttons || []).map(({ label, disabled, ...restProps }) => (
        <Button
            key={label}
            disabled={typeof disabled === 'function' ? disabled(params) : disabled}
            {...restProps}
        >
            {label}
        </Button>
    )) as Exclude<React.ReactNode, null | undefined | boolean | React.ReactFragment>;

export default ButtonsBlock;
