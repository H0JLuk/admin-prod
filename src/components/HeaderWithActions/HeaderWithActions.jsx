import React, { useCallback, useMemo } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { Button, Input } from 'antd';
import { SearchOutlined } from '@ant-design/icons';
import DropdownWithAction from '../DropdownWithAction/DropdownWithAction';

import { ReactComponent as Cross } from '../../static/images/cross.svg';

import styles from './HeaderWithActions.module.css';

const BUTTON_SORT = 'Сортировать';

const HeaderWithActions = ({
    title,
    buttons,
    searchInput,
    showSearchInput,
    showSorting,
    sortingBy,
    classNameByInput,
    classNames,
}) => {
    const buttonsBlock = useMemo(() => (buttons || []).map((button) => (
        <Button
            className={ cn(styles.button, { [styles.btn_primary]: button.type === 'primary' }) }
            key={ button.label }
            type={ button.type }
            onClick={ button.onClick }
            disabled={ button.disabled }
        >
            { button.label }
        </Button>
    )), [buttons]);

    const onChangeInput = useCallback(({ target: { value } }) => {
        searchInput.onChange(value);
    }, [searchInput]);

    const clickClearButton = useCallback(() => searchInput.onChange(''), [searchInput]);

    const dropdownLabel = useMemo(
        () => sortingBy.menuItems.find((item) => item.name === sortingBy.sortBy)?.label || BUTTON_SORT,
        [sortingBy.menuItems, sortingBy.sortBy]
    );

    return (
        <div className={ cn(styles.wrapper, classNames) }>
            {title && <h1>{ title }</h1>}
            <div className={ styles.actions }>
                <div className={ styles.buttons }>
                    { buttonsBlock }
                    {searchInput && showSearchInput && (
                        <Input
                            className={ cn(styles.searchInput, classNameByInput) }
                            onChange={ onChangeInput }
                            placeholder={ searchInput.placeholder }
                            prefix={ <SearchOutlined /> }
                            value={ searchInput.value }
                            suffix={ searchInput.value !== '' ?
                                <Cross
                                    className={ styles.crossIcon }
                                    onClick={ clickClearButton }
                                /> :
                                <span />
                            }
                            disabled={ searchInput.disabled }
                        />
                    )}
                </div>
                {sortingBy && showSorting && (
                    <DropdownWithAction
                        dropdownLabel={ dropdownLabel }
                        menuItems={ sortingBy.menuItems }
                        onMenuItemClick={ sortingBy.onMenuItemClick }
                        withReset={ sortingBy.withReset }
                    />
                )}
            </div>
        </div>
    );
};

HeaderWithActions.propTypes = {
    title: PropTypes.string,
    buttons: PropTypes.arrayOf(PropTypes.shape({
        type: PropTypes.string,
        label: PropTypes.string.isRequired,
        onClick: PropTypes.func.isRequired,
        disabled: PropTypes.bool,
    })),
    searchInput: PropTypes.shape({
        placeholder: PropTypes.string,
        onChange: PropTypes.func,
        disabled: PropTypes.bool,
    }),
    showSorting: PropTypes.bool,
    sortingBy: PropTypes.shape({
        buttonLabel: PropTypes.string,
        onMenuItemClick: PropTypes.func.isRequired,
        menuItems: PropTypes.arrayOf(PropTypes.shape({
            name: PropTypes.string.isRequired,
            label: PropTypes.string.isRequired,
            active: PropTypes.bool,
        })).isRequired,
        sortBy: PropTypes.string,
        withReset: PropTypes.bool,
    }),
};

export default HeaderWithActions;
