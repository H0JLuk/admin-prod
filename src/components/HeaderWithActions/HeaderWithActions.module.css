.wrapper {
    padding: 20px;
    background-color: #ffffff;
    letter-spacing: -0.3px;
}

.headerTitle {
    font-family: SBSansText;
    font-size: 20px;
    font-weight: 600;
    line-height: 1.4;
    letter-spacing: -0.38px;
    color: #171725;
    margin-bottom: 16px;
}

.actions {
    display: flex;
    justify-content: space-between;
    flex-wrap: nowrap;
}

.crossIcon {
    width: 22px;
    height: 22px;
    cursor: pointer;
}

.buttons {
    display: flex;
}

.buttons > button {
    margin-right: 10px;
}

.buttons > :global(.ant-input-affix-wrapper) {
    min-width: 250px;
}

.buttons > :global(.ant-input-affix-wrapper) span[role='img'] {
    color: #c2c8d2;
}
