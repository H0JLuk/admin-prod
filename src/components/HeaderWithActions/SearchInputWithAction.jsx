import React, { memo } from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';
import { Input } from 'antd';
import { SearchOutlined } from '@ant-design/icons';

const SearchInputWithAction = ({
    placeholder,
    onChangeValue,
    params,
    onChangeInput = noop,
}) => {
    const onChangeHandler = ({ target: { value } }) => {
        onChangeInput();
        onChangeValue({ ...params, filterText: value });
    };

    return (
        <Input
            onChange={ onChangeHandler }
            placeholder={ placeholder }
            value={ params.filterText }
            prefix={ <SearchOutlined /> }
            allowClear
        />
    );
};

SearchInputWithAction.propTypes = {
    placeholder: PropTypes.string.isRequired,
    onChangeValue: PropTypes.func.isRequired,
    params: PropTypes.object.isRequired,
    onChangeInput: PropTypes.func
};

export default memo(SearchInputWithAction);
