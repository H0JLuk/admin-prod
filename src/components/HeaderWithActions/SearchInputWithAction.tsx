import React, { ChangeEvent, memo } from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';
import { Input } from 'antd';
import { SearchOutlined } from '@ant-design/icons';
import { SearchParams } from './types';

type SearchInputWithActionProps = {
    placeholder: string;
    onChangeValue: (val: SearchParams) => void;
    params: SearchParams;
    onChangeInput?: () => void;
    resetPage?: boolean;
};

const SearchInputWithAction: React.FC<SearchInputWithActionProps> = ({
    placeholder,
    onChangeValue,
    params,
    onChangeInput = noop,
    resetPage,
}) => {
    const onChangeHandler = ({ target: { value } }: ChangeEvent<HTMLInputElement>) => {
        onChangeInput();
        resetPage && (params.pageNo = 0);
        onChangeValue({ ...params, filterText: value });
    };

    return (
        <Input
            onChange={onChangeHandler}
            placeholder={placeholder}
            value={params.filterText}
            prefix={<SearchOutlined />}
            allowClear
        />
    );
};

SearchInputWithAction.propTypes = {
    placeholder: PropTypes.string.isRequired,
    onChangeValue: PropTypes.func.isRequired,
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    params: PropTypes.object.isRequired,
    onChangeInput: PropTypes.func,
};

export default memo(SearchInputWithAction);
