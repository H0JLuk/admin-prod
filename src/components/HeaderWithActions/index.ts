import HeaderWithActions from './HeaderWithActions';

export type { ButtonProps } from './ButtonsBlock';

export default HeaderWithActions;
