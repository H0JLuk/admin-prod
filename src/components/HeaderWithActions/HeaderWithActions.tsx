import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import { useHistory } from 'react-router-dom';
import debounce from 'lodash/debounce';
import noop from 'lodash/noop';
import { DIRECTION } from '@constants/common';
import ButtonsBlock from './ButtonsBlock';
import SearchInputWithAction from './SearchInputWithAction';
import DropdownWithAction from '../DropdownWithAction/DropdownWithAction';
import {
    defaultSearchParams,
    getURLSearchParams,
    sortItemsBySearchParams,
    defaultPaginationParams,
    getSearchParamsFromUrl,
} from '@utils/helper';
import { HeaderWithActionsProps, SearchParams } from './types';

import styles from './HeaderWithActions.module.css';

const BUTTON_SORT = 'Сортировать';
const EMPTY_OBJECT: any = {};
const EMPTY_ARRAY: any[] = [];

function HeaderWithActions<DataList>({
    title,
    buttons,
    showSearchInput = true,
    showSorting = true,
    triggerResetParams,
    setDataList = noop,
    copyDataList = EMPTY_ARRAY,
    matchPath,
    sortByFieldKey,
    menuItems=[],
    inputPlaceholder,
    resetLabel,
    params = EMPTY_OBJECT,
    setParams = noop,
    onChangeInput,
    onChangeSort = noop,
    loading,
    loadData = noop,
    enableAsyncSort,
    enableHistoryReplace = true,
}: HeaderWithActionsProps<DataList>) {
    const history = useHistory();
    const { search } = history.location;
    const [paramsSync, setParamsSync] = useState<SearchParams>(getSearchParamsFromUrl(search));
    const initialRender = useRef(true);

    useEffect(() => {
        if (!enableAsyncSort) {
            setSyncParams(paramsSync);
        }
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [copyDataList]);

    useEffect(() => {
        if (!enableAsyncSort) {
            if (!initialRender.current) {
                setSyncParams(defaultSearchParams);
            } else {
                initialRender.current = false;
            }
        }
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [triggerResetParams]);

    // eslint-disable-next-line react-hooks/exhaustive-deps
    const loadDataDebounced = useCallback(debounce(loadData, 500), [loadData]);

    const currentParams = enableAsyncSort ? params : paramsSync;

    const setAsyncParams = useCallback((searchParams: SearchParams = defaultPaginationParams) => {
        if (loading) {
            return;
        }

        if (searchParams.filterText) {
            loadDataDebounced(searchParams);
        } else {
            loadDataDebounced.cancel();
            loadData(searchParams);
        }

        setParams(searchParams);
    }, [loadData, loadDataDebounced, loading, setParams]);

    const setSyncParams = useCallback((searchParams: SearchParams = defaultSearchParams) => {
        setParamsSync(searchParams);
        enableHistoryReplace && history.replace(`${matchPath}?${getURLSearchParams(searchParams)}`);

        if (!searchParams.filterText && !searchParams.sortBy) {
            setDataList(copyDataList);
            return;
        }

        setDataList(sortItemsBySearchParams(searchParams, copyDataList, sortByFieldKey!));
    }, [copyDataList, enableHistoryReplace, history, matchPath, setDataList, sortByFieldKey]);

    const sortData = enableAsyncSort ? setAsyncParams : setSyncParams;

    const changeSort = useCallback((sortBy) => {
        onChangeSort();

        if (typeof sortBy !== 'string') {
            sortBy = '';
        }

        if (!sortBy && !currentParams.sortBy) return;

        sortData({
            ...currentParams,
            sortBy,
            direction: !sortBy || currentParams.direction === DIRECTION.DESC ? DIRECTION.ASC : DIRECTION.DESC,
        });
    }, [currentParams, onChangeSort, sortData]);

    const dropdownLabel = useMemo(() => {
        const { label } = menuItems.find((item) => item.name === currentParams.sortBy) || {};
        return label || BUTTON_SORT;
    }, [menuItems, currentParams.sortBy]);

    const withReset = currentParams.sortBy !== '';

    return (
        <div className={styles.wrapper}>
            {title && <div className={styles.headerTitle}>{title}</div>}
            <div className={styles.actions}>
                <div className={styles.buttons}>
                    <ButtonsBlock
                        buttons={buttons}
                        params={currentParams}
                    />
                    {showSearchInput && (
                        <SearchInputWithAction
                            placeholder={inputPlaceholder}
                            onChangeValue={sortData}
                            params={currentParams}
                            onChangeInput={onChangeInput}
                            resetPage={enableAsyncSort}
                        />
                    )}
                </div>
                {showSorting && (
                    <DropdownWithAction
                        dropdownLabel={dropdownLabel}
                        menuItems={menuItems}
                        onMenuItemClick={changeSort}
                        resetLabel={resetLabel}
                        withReset={withReset}
                    />
                )}
            </div>
        </div>
    );
}

HeaderWithActions.propTypes = {
    title: PropTypes.string,
    buttons: PropTypes.arrayOf(PropTypes.shape({
        type: PropTypes.string,
        label: PropTypes.string.isRequired,
        onClick: PropTypes.func.isRequired,
        disabled: PropTypes.oneOfType([
            PropTypes.bool,
            PropTypes.func,
        ]),
    })),
    showSearchInput: PropTypes.bool,
    showSorting: PropTypes.bool,
    triggerResetParams: PropTypes.any,
    setDataList: PropTypes.func,
    copyDataList: PropTypes.array,
    matchPath: PropTypes.string,
    sortByFieldKey: PropTypes.string,
    menuItems: PropTypes.array,
    inputPlaceholder: PropTypes.string,
    resetLabel: PropTypes.string,
    params: PropTypes.object,
    setParams: PropTypes.func,
    onChangeInput: PropTypes.func,
    onChangeSort: PropTypes.func,
    loading: PropTypes.bool,
    loadData: PropTypes.func,
    enableAsyncSort: PropTypes.bool,
    enableHistoryReplace: PropTypes.bool,
};

export default HeaderWithActions;
