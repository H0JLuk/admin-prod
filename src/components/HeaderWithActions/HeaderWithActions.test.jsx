import React from 'react';
import { mount, shallow } from 'enzyme';
import HeaderWithActions from './HeaderWithActions';
import { sleep } from '../../setupTests';
import { defaultSearchParams } from '../../utils/helper';

const TEST_BUTTONS_PROPS = [
    {
        label: 'TEST_TEXT_ONE',
        type: 'TEST_PRIMARY_TYPE',
        onClick: jest.fn(),
        disabled: false,
    },
    {
        label: 'TEST_TEXT_TWO',
        type: 'TEST_DANGER_TYPE',
        onClick: jest.fn(),
        disabled: true,
    },
];

const TEST_MENU_ITEMS = [
    {
        name: 'TEST_NAME_ONE',
        label: 'TEST_LABEL_ONE',
        active: false
    },
    {
        name: 'TEST_NAME_TWO',
        label: 'TEST_LABEL_TWO',
        active: false
    }
];

const TEST_SYNC_PROPS = {
    title: 'TEST_TITLE_PROP',
    buttons: TEST_BUTTONS_PROPS,
    showSearchInput: true,
    showSorting: true,
    triggerResetParams: false,
    setDataList: jest.fn(),
    copyDataList: [],
    matchPath: 'TEST_PATH',
    sortByFieldKey: 'TEST_KEY',
    menuItems: TEST_MENU_ITEMS,
    inputPlaceholder: 'TEST_PLACEHOLDER',
    resetLabel: 'TEST_LABEL',
    params: defaultSearchParams,
    setParams: jest.fn(),
    onChangeInput: jest.fn(),
    onChangeSort: jest.fn(),
    loading: false,
    loadData: jest.fn(),
    enableAsyncSort: false,
    enableHistoryReplace: true,
};

const TEST_ASYNC_PROPS = { ...TEST_SYNC_PROPS, enableAsyncSort: true };

const TEST_VALUE = 'TEST_VALUE';

const TEST_SORT_BY = {
    STRING: 'TEST',
    NUMBER: 123,
};

const mock = jest.fn();

jest.mock('react-router-dom', () => ({
    useParams: () => ({ userId: 1 }),
    useHistory: () => ({
        replace: mock,
        location: {
            search: 'pageNo=1&pageSize=20&sortBy=&direction=ASC&filterText=',
        },
    }),
    generatePath: jest.fn()
}));

describe('<HeaderWithActions /> tests.', () => {
    const Header = mount(
        <HeaderWithActions { ...TEST_SYNC_PROPS } />
    );

    it('should be mount', () => {
        expect(Header.html()).toMatchSnapshot();
    });

    it('should render title from props', () => {
        expect(Header.find('div.headerTitle').text()).toBe(TEST_SYNC_PROPS.title);
    });

    it('should call change input function', () => {
        Header.find('input').simulate('change', { target: { value: TEST_VALUE } });
        expect(TEST_SYNC_PROPS.onChangeInput).toHaveBeenCalledTimes(1);
    });

    it('should call change input function with async params', async () => {
        const HeaderAsync = mount(
            <HeaderWithActions { ...TEST_ASYNC_PROPS } />
        );
        HeaderAsync.find('input').simulate('change', { target: { value: TEST_VALUE } });
        // sleep need because loadData function is debounced
        await sleep(501);
        expect(TEST_SYNC_PROPS.loadData).toHaveBeenCalledTimes(1);
    });

    it('should call `changeSort` function', () => {
        const Header = shallow(
            <HeaderWithActions { ...TEST_SYNC_PROPS } />
        );
        const DropDown = Header.findWhere(n => n.name() === 'Memo(DropdownWithAction)');
        DropDown.props().onMenuItemClick(TEST_SORT_BY.STRING);
        expect(TEST_SYNC_PROPS.onChangeSort).toBeCalledTimes(1);
        expect(mock).toBeCalledTimes(1);
    });

    it('should call `changeSort` function and clear sortBy if sortBy !== string', () => {
        const Header = shallow(
            <HeaderWithActions { ...TEST_SYNC_PROPS } />
        );
        const DropDown = Header.find('Memo(DropdownWithAction)');
        DropDown.props().onMenuItemClick(TEST_SORT_BY.NUMBER);
        expect(TEST_SYNC_PROPS.onChangeSort).toBeCalledTimes(1);
        expect(mock).not.toBeCalledTimes(1);
    });

    it('should call debounce with async props', async () => {
        const Header = shallow(
            <HeaderWithActions { ...TEST_ASYNC_PROPS } />
        );
        const SearchInput = Header.find('Memo(SearchInputWithAction)');
        SearchInput.props().onChangeValue();
        await sleep(501);
        expect(TEST_SYNC_PROPS.loadData).toHaveBeenCalledTimes(1);
    });

    it('should not call debounce with async props', async () => {
        const Header = shallow(
            <HeaderWithActions { ...TEST_ASYNC_PROPS } loading={ true } />
        );
        const SearchInput = Header.find('Memo(SearchInputWithAction)');
        SearchInput.props().onChangeValue();
        expect(TEST_SYNC_PROPS.loadData).toHaveBeenCalledTimes(0);
    });

    it('should rerender if change triggerParams', () => {
        const SearchInput = Header.find('Memo(SearchInputWithAction)');
        expect(SearchInput.prop('params')).toEqual({ 'direction': 'ASC', 'filterText': 'TEST_VALUE', 'sortBy': '' });
        Header.setProps({ triggerResetParams: true });
        Header.update();
        expect(Header.find('Memo(SearchInputWithAction)').prop('params')).toEqual(defaultSearchParams);
    });

});
