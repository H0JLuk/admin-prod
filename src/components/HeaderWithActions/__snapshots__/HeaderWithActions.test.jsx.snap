// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderWithActions /> tests. should be mount 1`] = `"<div class=\\"wrapper\\"><div class=\\"headerTitle\\">TEST_TITLE_PROP</div><div class=\\"actions\\"><div class=\\"buttons\\"><button type=\\"button\\" class=\\"ant-btn ant-btn-TEST_PRIMARY_TYPE\\"><span>TEST_TEXT_ONE</span></button><button disabled=\\"\\" type=\\"button\\" class=\\"ant-btn ant-btn-TEST_DANGER_TYPE\\"><span>TEST_TEXT_TWO</span></button><span class=\\"ant-input-affix-wrapper\\"><span class=\\"ant-input-prefix\\"><span role=\\"img\\" aria-label=\\"search\\" class=\\"anticon anticon-search\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"search\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z\\"></path></svg></span></span><input placeholder=\\"TEST_PLACEHOLDER\\" type=\\"text\\" class=\\"ant-input\\" value=\\"\\"><span class=\\"ant-input-suffix\\"><span role=\\"button\\" aria-label=\\"close-circle\\" tabindex=\\"-1\\" class=\\"anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"close-circle\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z\\"></path></svg></span></span></span></div><button type=\\"button\\" class=\\"ant-btn ant-dropdown-trigger dropdownButton\\"><span>Сортировать </span><span role=\\"img\\" aria-label=\\"down\\" style=\\"font-size: 14px; color: rgb(195, 201, 208);\\" class=\\"anticon anticon-down\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></span></button></div></div>"`;
